src/libdredd/src/mutation_replace_binary_operator.cc

```diff
@@ -1024,34 +1029,108 @@ MutationReplaceBinaryOperator::ClangOperatorKindToProtobufOperatorKind(
+                                                  ast_context)) {
+    return true;
+  }
+
   switch (binary_operator_->getOpcode()) {
     // From
     // https://people.cs.umass.edu/~rjust/publ/non_redundant_mutants_jstvr_2014.pdf:
     // For boolean operators, only a subset of replacements are non-redundant.
     case clang::BO_LAnd:
-      return operator_kind != clang::BO_EQ;
+      return replacement_operator != clang::BO_EQ;
     case clang::BO_LOr:
-      return operator_kind != clang::BO_NE;
+      return replacement_operator != clang::BO_NE;
     case clang::BO_GT:
-      return operator_kind != clang::BO_GE && operator_kind != clang::BO_NE;
+      return replacement_operator != clang::BO_GE &&
+             replacement_operator != clang::BO_NE;
     case clang::BO_GE:
-      return operator_kind != clang::BO_GT && operator_kind != clang::BO_EQ;
+      return replacement_operator != clang::BO_GT &&
+             replacement_operator != clang::BO_EQ;
     case clang::BO_LT:
-      return operator_kind != clang::BO_LE && operator_kind != clang::BO_NE;
+      return replacement_operator != clang::BO_LE &&
+             replacement_operator != clang::BO_NE;
     case clang::BO_LE:
-      return operator_kind != clang::BO_LT && operator_kind != clang::BO_EQ;
+      return replacement_operator != clang::BO_LT &&
+             replacement_operator != clang::BO_EQ;
     case clang::BO_EQ:
-      return operator_kind != clang::BO_LE && operator_kind != clang::BO_GE;
+      return replacement_operator != clang::BO_LE &&
+             replacement_operator != clang::BO_GE;
     case clang::BO_NE:
-      return operator_kind != clang::BO_LT && operator_kind != clang::BO_GT;
+      return replacement_operator != clang::BO_LT &&
+             replacement_operator != clang::BO_GT;
     default:
       return false;
   }
 }
 
+bool MutationReplaceBinaryOperator::IsRedundantReplacementForUnsignedComparison(
+    clang::BinaryOperatorKind replacement_operator,
+    const clang::ASTContext& ast_context) const {
+  if (!binary_operator_->getLHS()->getType()->isUnsignedIntegerType() ||
+      !binary_operator_->getRHS()->getType()->isUnsignedIntegerType()) {
+    return false;
+  }
+  if (MutationReplaceExpr::ExprIsEquivalentToInt(*binary_operator_->getLHS(), 0,
+                                                 ast_context)) {
+    // LHS is 0
+    if (replacement_operator == clang::BO_GT ||
+        replacement_operator == clang::BO_LE) {
+      // We are considering mutating an expression of the form "0 op a" to
+      // one of:
+      //
+      // - "0 > a"
+      // - "0 <= a"
+      //
+      // These are subsumed by replacing the overall expression with "false"
+      // and "true", respectively.
+      return true;
+    }
+    if (std::set({binary_operator_->getOpcode(), replacement_operator}) ==
+        std::set<clang::BinaryOperatorKind>({clang::BO_EQ, clang::BO_GE})) {
+      // "0 == a" is equivalent to "0 >= a"
+      return true;
+    }
```

Comment:
By the fact that `0 == a` is equivalent to `0 >= a` and commutativity, it is also redundant to replace `0 >= a` by `0 == a` for signed integers (which is unlikely but may occur). I think equality on ordered sets compare element wise so this wouldn't currently cover the second case.